### ISSUE TITLE
(II) Question on GLOSA code

### ISSUE THREAD

**Issue by thinker310** [eclipse-sumo/sumo#12217]
Hi @namdre, thank you for your explanation, I have some questions in detail and hope to your further comments.
1.Continue to post #12211, in my opinion, there should be two root_argument variables, one for green light and another for red variable. Meanwhile, the meaning of this formula is confusing, such as _a*d_ in the formula, why acceleration multiplyies by deceleration?

2.For the annotations on line 288 and 291, _y_ is slow down duration, _y_ can be _(w-x)/d_, however _y=t-(w-x)/d_ on line 291. Also, on line 290, the second term _x*(t-(w-x)/d-(v-x)/d)_ is understandable but still a little mistake, for acceleration segmentation, denominator should be _a_, for deceleration segmentation, denominator should be _d_. In the annotation code, both denominators used _d_. 

3.On line 289 and 290, there exist more than one scenario that a car approaches to the intersection. However, only one situation was considered as the description on line 289. 

4.What is the role of _sign0_?

Looking forward to your reply. Thank you again!

**Comment by namdre** [comment]
1. `adaptSpeed` only solves for the case where the light is red 
2. I agree that there is an inconsistency here. It should be `y = t - (w-x) / a` (because  `w-x` concerns the acceleration to maximum speed after having slowed down from v to x earlier.
3. No. The scenario is always the same here: decelerate, then keep low speed for some time, finally accelerate back to maximum speed. However, some of these components may have zero duration. Arguably, we may have to consider x > v and the re-solve with 'a' instead of 'd' for the first part of the sequence.
4. Generally speaking, quadratic formulas have two solutions so the sign selects which one of these is used

**Comment by thinker310** [comment]
Ok, thank you! Furthermore, I wonder how to implement the target speed to make cars **track GLOSA**, need I use `traci.vehicle.slowDown(targetSpeed, duration)` for **every simulation step**?

**Comment by thinker310** [comment]
Ok, thank you! Furthermore, I wonder how to implement the target speed to make cars **track GLOSA**, need I use `traci.vehicle.slowDown(targetSpeed, duration)` for **every simulation step**?

**Comment by namdre** [comment]
the 'slowDown' command takes effect over the given 'duration' argument so if the speed advice does not change, you only need to send a new command after 'duration'.

**Comment by thinker310** [comment]
yeah, but for coding, it is hard to implement this logic as shown in the picture below:
![uTools_1670423868030](https://user-images.githubusercontent.com/62609436/206208246-6287bd03-1c87-498d-8650-476e86b309aa.png)
In the simulation loop, every loop the `traci.vehicle.slowDown` command will be executed, how to execute speed command after 'duration'?

**Comment by namdre** [comment]
Upon closer inspection, the computation of the slow-down duration (`y`) neglects the time at which the speed must be maintained and thus results in insufficient deceleration.

**Comment by behrisch** [comment]
here is how we changed the computation of the remaining duration in a first ad hoc patch (line 131 of aglosa_common.py):
```
    if x < speed:
        y = (speed - x) / d
    else:
        y = (x - speed) / a
```

**Comment by thinker310** [comment]
> here is how we changed the computation of the remaining duration in a first ad hoc patch (line 131 of aglosa_common.py):
> 
> ```
>     if x < speed:
>         y = (speed - x) / d
>     else:
>         y = (x - speed) / a
> ```

Could you share aglosa_common.py file with me?  I did not find in the repository.

**Comment by thinker310** [comment]
Hello, there exist some problems when I implement speed guidance strategy, could you provide some details on 'track_glosa(id, target_speed, duration)' and 'track_glosa_end(id)' please?

**Comment by namdre** [comment]
both functions are only for printing debug output and do not affect vehicle behavior.

**Comment by namdre** [comment]
fixed via #15067
